pandas/core/algorithms.py

```diff
@@ -944,14 +944,64 @@ def select_n_frame(frame, columns, n, method, keep):
     -------
     nordered : DataFrame
     """
-    from pandas.core.series import Series
+    from pandas import Int64Index
     if not is_list_like(columns):
         columns = [columns]
     columns = list(columns)
-    ser = getattr(frame[columns[0]], method)(n, keep=keep)
-    if isinstance(ser, Series):
-        ser = ser.to_frame()
-    return ser.merge(frame, on=columns[0], left_index=True)[frame.columns]
+    for column in columns:
+        dtype = frame[column].dtype
+        if not issubclass(dtype.type, (np.integer, np.floating, np.datetime64,
```

Comment:
Ah right, ok, sorry I thought by your "check upfront" comment you wanted to do it for each column in this method.
But you're right the line ``getattr(series, method)(cur_n, keep=keep)`` should defer the check to n* on the series